src/interpreter/interpreter-non-det.ts

```diff
@@ -436,14 +436,18 @@ export const evaluators: { [nodeType: string]: Evaluator<es.Node> } = {
   },
 
   BlockStatement: function*(node: es.BlockStatement, context: Context) {
-    let result: Value
-
     // Create a new environment (block scoping)
-    const environment = createBlockEnvironment(context, 'blockEnvironment')
+    let environment = createBlockEnvironment(context, 'blockEnvironment')
     pushEnvironment(context, environment)
-    result = yield* evaluateBlockSatement(context, node)
+    const resultGenerator = evaluateBlockSatement(context, node)
+
+    for (const result of resultGenerator) {
+      popEnvironment(context)
+      yield result
+      environment = createBlockEnvironment(context, 'blockEnvironment')
```

Comment:
Shouldn't we re-use the same environment created in line 440?

Reply:
Yes that should be so.